mifosng-provider/src/main/java/org/mifosplatform/portfolio/loanaccount/domain/Loan.java

```diff
@@ -5615,6 +5640,75 @@ public void addTrancheLoanCharge(Charge charge) {
             trancheCharges.add(new LoanTrancheCharge(charge, this));
         }
     }
+    
+    public List<LoanTransaction> undoLastDisbursal(ScheduleGeneratorDTO scheduleGeneratorDTO, List<Long> existingTransactionIds,
+			List<Long> existingReversedTransactionIds, AppUser currentUser, Map<String, Object> changes, Loan loan) {
+
+
+		validateAccountStatus(LoanEvent.LOAN_DISBURSAL_UNDOLAST);
+
+        List<LoanTransaction> reversedTransactions = new ArrayList<>();
+        validateActivityNotBeforeClientOrGroupTransferDate(LoanEvent.LOAN_DISBURSAL_UNDOLAST, getDisbursementDate());
+        LocalDate actualDisbursementDate = null;
+    	LocalDate lastTransactionDate = getDisbursementDate();
+    	List<LoanTransaction> loanTransactions = retreiveListOfTransactionsExcludeAccruals();
+    	Collections.reverse(loanTransactions);
+    	for (final LoanTransaction previousTransaction : loanTransactions) {
+            if (!(previousTransaction.isWaiver())) {
+                if (lastTransactionDate.isBefore(previousTransaction.getTransactionDate())) {
+                	if(previousTransaction.isRepayment()){
+                		throw new UndoLastTrancheDisbursementException(previousTransaction.getId()); }
+                	}
+                lastTransactionDate = previousTransaction.getTransactionDate();
+                break;
+            }
+        }
+        actualDisbursementDate = lastTransactionDate;
+        updateLoanToLastDisbursalState(actualDisbursementDate);
+        reversedTransactions = reverseExistingTransactionsTillLastDisbursal(actualDisbursementDate);
+        loan.recalculateScheduleFromLastTransaction(scheduleGeneratorDTO,
+                existingTransactionIds, existingReversedTransactionIds, currentUser);
+        changes.put("undolastdisbursal","true");
+        existingTransactionIds.addAll(findExistingTransactionIds());
```

Comment:
Why you are adding reversed transactions after some of the transactions reversed. This should the first statement. Otherwise you accounting won't work properly